godot-core/src/builtin/dictionary.rs

```diff
@@ -0,0 +1,314 @@
+            (interface_fn!(dictionary_operator_index_const))(self.sys_const(), key.var_sys_const()) as *const Variant
+        }
+    }
+
+    /// Get the pointer corresponding to the given key in the dictionary,
+    /// if there exists no value at the given key then a new one is created
+    /// and initialized to a nil variant.
+    pub fn get_ptr_mut<K: ToVariant>(&mut self, key: K) -> *mut Variant {
+        let key = key.to_variant();
+        unsafe {
+            let ptr =
+                (interface_fn!(dictionary_operator_index))(self.sys_mut(), key.var_sys_const()) as *mut Variant;
+            // dictionary_operator_index initializes the value so it wont be null
+            assert!(!ptr.is_null());
+            // i think it might be safe to turn this into a &mut Variant?
+            ptr
+        }
+    }
+
+    /// Insert a value at the given key, returning the value
+    /// that previously was at that key if there was one.
+    pub fn insert<K: ToVariant>(&mut self, key: K, value: Variant) -> Option<Variant> {
+        let key = key.to_variant();
+        let old_value = self.remove(key.clone());
+        self.set(key, value);
+        old_value
+    }
+
+    /// Set a key to a given value
+    pub fn set<K: ToVariant>(&mut self, key: K, value: Variant) {
+        let key = key.to_variant();
+        unsafe {
+            *self.get_ptr_mut(key) = value;
+        }
+    }
+
+    /// Convert this dictionary to a strongly typed rust `HashMap`. If the conversion
+    /// fails for any key or value, an error is returned. 
+    pub fn try_to_hashmap<K: FromVariant + Eq + std::hash::Hash, V: FromVariant>(
+        &self,
+    ) -> Result<HashMap<K, V>, VariantConversionError> {
+        let mut map = HashMap::new();
+        for key in self.keys().into_iter() {
+            map.insert(key.try_to()?, self.get(key).unwrap().try_to()?);
+        }
+        Ok(map)
+    }
+
+    /// Convert the keys of this dictionary to a strongly typed rust `HashSet`. If the 
+    /// conversion fails for any key, an error is returned. 
+    pub fn try_to_hashset<K: FromVariant + Eq + std::hash::Hash>(
+        &self,
+    ) -> Result<HashSet<K>, VariantConversionError> {
+        Ok(self.keys().try_to_vec::<K>()?.into_iter().collect())
+    }
+
+    #[doc(hidden)]
+    pub fn as_inner(&self) -> inner::InnerDictionary {
+        inner::InnerDictionary::from_outer(self)
+    }
+}
+
+/// Creates a `Dictionary` from the given `HashMap`. Each key and value are 
+/// converted to a `Variant`.
+impl<K: ToVariant, V: ToVariant> From<HashMap<K, V>> for Dictionary {
+    fn from(map: HashMap<K, V>) -> Self {
+        let mut dict = Dictionary::new();
+        for (k, v) in map.into_iter() {
+            dict.insert(k.to_variant(), v.to_variant());
+        }
+        dict
+    }
+}
+
+/// Creates a `Dictionary` from the given `HashSet`. Each key is converted
+/// to a `Variant`, and the values are all `true`.
+impl<K: ToVariant> From<HashSet<K>> for Dictionary {
```

Comment:
Maybe this is a bit specific. Why `true` and not `0` or `nil`?

A `From<T> where T: IntoIterator<(K, V)>` (same as suggestion above, but taking `K` and `V` by value) would cover this use case (among others):

```
my_hash_set.into_iter().map(|k| (k, true)).collect::<Dictionary>()
```

Reply:
Good point, I don't think there should be a direct conversion from sets to maps without specifying the behavior for values.

Reply:
`true` is more useful because it allows:

```gdscript
if dict_set[key]:
    do_stuff()
```

...which might feel a bit more natural for GDScript users, instead of the *slightly* more verbose:

```gdscript
if dict_set.has(key):
    do_stuff()
```

Nevertheless, I agree that it's better to leave the exact behavior up to users, and not implement direct conversion from sets.